model/LtiListenerService.php

```diff
@@ -40,31 +43,73 @@ public function executionCreated(DeliveryExecutionCreated $event)
                 $tagsString = implode(',', $tags);
                 $tagsString = str_pad($tagsString, strlen($tagsString) + 2, ',', STR_PAD_BOTH);
             }
-            $resourceLink = $session->getLaunchData()->getResourceLinkID();
-            $this->getServiceLocator()->get(DeliveryLog::SERVICE_ID)->log(
-                $event->getDeliveryExecution()->getIdentifier(), 'LTI_DELIVERY_EXECUTION_CREATED', [
-                    LtiLaunchData::CONTEXT_ID => $contextId,
-                    LtiLaunchData::CONTEXT_LABEL => $session->getLaunchData()->getVariable(LtiLaunchData::CONTEXT_LABEL),
-                    LtiLaunchData::RESOURCE_LINK_ID => $resourceLink,
-                    ProctorService::CUSTOM_TAG => $tagsString,
-                ]
-            );
-            $monitoringService = $this->getServiceManager()->get(DeliveryMonitoringService::SERVICE_ID);
-            $data = $monitoringService->getData($event->getDeliveryExecution());
+            $resourceLink = $launchData->getResourceLinkID();
+            $deliveryLog->log($executionId, 'LTI_DELIVERY_EXECUTION_CREATED', [
+                LtiLaunchData::CONTEXT_ID => $contextId,
+                LtiLaunchData::CONTEXT_LABEL => $launchData->getVariable(LtiLaunchData::CONTEXT_LABEL),
+                LtiLaunchData::RESOURCE_LINK_ID => $resourceLink,
+                ProctorService::CUSTOM_TAG => $tagsString,
+            ]);
+
+            $monitoringService = $serviceManager->get(DeliveryMonitoringService::SERVICE_ID);
+            $data = $monitoringService->getData($deliveryExecution);
             $data->update(LtiLaunchData::CONTEXT_ID, $contextId);
             $data->update(LtiLaunchData::RESOURCE_LINK_ID, $resourceLink);
             $data->update(ProctorService::CUSTOM_TAG, $tagsString);
+
+            if ($launchData->hasVariable(LtiDeliveryExecutionService::LTI_USER_NAME)) {
+                $ltiUserName = $launchData->getVariable(LtiDeliveryExecutionService::LTI_USER_NAME);
+                $deliveryLog->log($executionId, 'LTI_USER_NAME', $ltiUserName);
+
+                $data->update(LtiDeliveryExecutionService::LTI_USER_NAME, $ltiUserName);
+            }
+
             $success = $monitoringService->save($data);
             if (!$success) {
-                \common_Logger::w('monitor cache for delivery ' . $event->getDeliveryExecution()->getIdentifier() . ' could not be created');
+                \common_Logger::w('monitor cache for delivery ' . $executionId . ' could not be created');
+            }
+        }
+    }
+
+    public function executionStateChanged(DeliveryExecutionState $event)
+    {
+        $session = \common_session_SessionManager::getSession();
+        if ($session instanceof \taoLti_models_classes_TaoLtiSession) {
+            $launchData = $session->getLaunchData();
+            if ($event->getState() == DeliveryExecution::STATE_ACTIVE &&
+                $event->getPreviousState() == DeliveryExecution::STATE_AUTHORIZED &&
```

Comment:
You are right. The session may not be proctored but a proctor could see it. So I removed this check.